src/map_memory.cpp

```diff
@@ -0,0 +1,59 @@
+#include "map_memory.h"
+
+#include <algorithm>
+
+void map_memory::trim( int limit )
+{
+    while( tiles.size() > static_cast<size_t>( limit ) ) {
+        tile_map.erase( tiles.back().first );
+        tiles.pop_back();
+    }
+    while( symbols.size() > static_cast<size_t>( limit ) ) {
+        symbol_map.erase( symbols.back().first );
+        symbols.pop_back();
+    }
+}
+
+memorized_terrain_tile map_memory::get_tile( const tripoint &pos ) const
+{
+    auto found_tile = tile_map.find( pos );
+    if( found_tile != tile_map.end() ) {
+        return found_tile->second->second;
+    }
+    return { "", 0, 0 };
+}
+
+void map_memory::memorize_tile( int limit, const tripoint &pos, const std::string &ter,
+                                const int subtile, const int rotation )
+{
+    memorized_terrain_tile new_tile{ ter, subtile, rotation };
+    tiles.push_front( std::make_pair( pos, new_tile ) );
+    auto found_tile = tile_map.find( pos );
+    if( found_tile != tile_map.end() ) {
+        // Remove redundant entry since we pushed one to the front.
+        tiles.erase( found_tile->second );
```

Comment:
Have you considered keeping `memorized_terrain_tile` in the map rather than in the list? This way it won't be deallocated on every memorization. You could even check that nothing's changed since the previous memorization and leave the map intact only updating the list. Tiles rarely change.